packages/apollo-gateway/src/FieldSet.ts

```diff
@@ -46,6 +46,11 @@ export function matchesField(field: Field) {
   };
 }
 
+export const isEmpty = (keyFields: FieldSet) => (
+  keyFields.length === 0 ||
+    (keyFields.length === 1 && keyFields[0].fieldDef.name === '__typename')
+);
+
```

Comment:
The name could be misleading. What do you think about making it local lambda in `splitSubfields` function?

Reply:
personally I hate these local lamda as it's hard to track especially in this project where each function is a monster with hundreds of LOC having multiple local lambdas which are called recursively mutation state and reusing variable from outer scope :)

By extracting these to separate function just prevents users from doing some bad things. Maybe it's a good idea to use another naming but I like `isEmpty`